Show Welcome Tour to all new users on Desktop, NUX to users on Mobile

autumnfjeld opened the pull request:
## Changes proposed in this Pull Request

- [x] Remove `apply_filter` code that hooked to experiment
- [x] Remove client code that handled the variant variable and determined which editor welcome to show to user
- [x] Remove  check fo Anchor.fm usage since  @allilevine has [REMOVED anchor NUX content](https://github.com/Automattic/wp-calypso/commit/1c2161f5e768d87353d42f9e4ad2ad3d36a06b36)
- [x] Show NUX modal to mobile users

A subsequent diff will remove the experiment code D53894-code

## Testing instructions

* pull down this branch
* `yarn dev --sync` to your sandbox
 OR
* from your sandbox: 
  * `cd /home/wpcom/public_html`  
  *  `install-plugin.sh etk update/show-welcome-tour-to-new-users-cept-anchorfm`  (nifty!)
* On a simple site that is sandboxed test the following:

* Clear application data:  In dev tools Application tab you must clear the site's local storage to ensure the @automattic/nux store is reset. Must be done in order to show the Welcome Guide on first load.

**Scenario #1 Desktop**
* Welcome Tour should show when opening editor
  * go through all slides, minimize, maximize, etc to make sure Tour works 
* Open Tour from Options Menu (aka More Menu) -> "Welcome Guide"
* Close, then re-open editor and make sure the Welcome Tour does not automatically show

**Scenario #2 Mobile**
* Clear application data agin
* Use Chrome dev tools to mimic mobile
 OR
* BETTER if you are using https://github.com/Automattic/a8c-wp-env you can easily test on your mobile device via the Jurassic Tube integration (or ngrok)
* The NUX Modal should show


Related to # https://github.com/Automattic/wp-calypso/issues/50877

autumnfjeld committed:
The welcome tour vs nux experiment is done: remove experiment filter and related code in the nux store and clientside launch code

autumnfjeld commented:
Hiya @allilevine 
This PR will enable showing the new [Welcome Tour](https://github.com/Automattic/wp-calypso/blob/49ff4ed6c48dfd79f8f90d09263ce97d90d8163f/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/block-editor-nux.js#L53) to all users.

I just now discovered we don't need to show the NUX modal for Anchor.fm users because you've [removed that code](https://github.com/Automattic/wp-calypso/commit/1c2161f5e768d87353d42f9e4ad2ad3d36a06b36).

In a subsequent PR we will delete the NUX code entirely, unless your team needs it for Anchor.fm content. However a much better solution would be to create new content in the Welcome Tour style.

allilevine commented:
> Hiya @allilevine
> This PR will enable showing the new [Welcome Tour](https://github.com/Automattic/wp-calypso/blob/49ff4ed6c48dfd79f8f90d09263ce97d90d8163f/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/block-editor-nux.js#L53) to all users.
> 
> I just now discovered we don't need to show the NUX modal for Anchor.fm users because you've [removed that code](https://github.com/Automattic/wp-calypso/commit/1c2161f5e768d87353d42f9e4ad2ad3d36a06b36).
> 
> In a subsequent PR we will delete the NUX code entirely, unless your team needs it for Anchor.fm content. However a much better solution would be to create new content in the Welcome Tour style.

@autumnfjeld Thanks for the heads-up! We removed the Anchor-specific Welcome Tour because it was no longer needed (we changed the first page Anchor signups edit to the homepage, instead of a podcast episode post). I don't think we need to continue to check for podcasting sites here, but cc-ing @kwight to double-check.

kwight commented:
@autumnfjeld @allilevine No, the Anchor project doesn't need anything outside of the Gutenboarding defaults – it should be considered the same as any other flow with regards to the Welcome Guide. Thanks for checking!

ramonjd commented:
@autumnfjeld 

Just checking that we're delivering this to all mobile users as well. 

We excluded them from the experiment. 

See: 28b76-pb

<img width="437" alt="Screen Shot 2021-03-16 at 9 57 46 am" src="https://user-images.githubusercontent.com/6458278/111231850-11d1f980-863e-11eb-8226-031fb0cbdb4e.png">

Cheers!

autumnfjeld commented:
@ramonjd 

> 
> Just checking that we're delivering this to all mobile users as well. 
> 
> We excluded them from the experiment. 


Thank you for this reminder!!  We do not want to show on mobile.  Ola is working on mobile designs, til then we show on desktop / larger screens only. Thus, we will need to identify mobile and prevent showing.  I will dig up the ideas we previously discussed to do this.

Cc @simison

ramonjd commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php:
```diff
@@ -69,10 +69,6 @@ public function is_nux_enabled( $nux_status ) {
 	public function get_nux_status() {
-		// Check if we want to show the Welcome Tour variant pbxNRc-Cb-p2
-		// Performing the check here means that we'll only assign a user to an experiment when this API is first called.
-		$welcome_tour_show_variant = ( defined( 'SHOW_WELCOME_TOUR' ) && SHOW_WELCOME_TOUR ) || apply_filters( 'a8c_enable_wpcom_welcome_tour', false );
```
> Thank you for this reminder!! We do not want to show on mobile. Ola is working on mobile designs, til then we show on desktop / larger screens only. Thus, we will need to identify mobile and prevent showing. 

Good to know. 

Once option might be to keep the filter here temporarily, since the mobile check is on the backend. 

Then, in the backend filter, we would remove the code related to the experiment.

Or we move the mobile check to the ETK (and remove the WPCOM filter entirely) so we have consistent behaviour across Simple and Atomic sites. What do you think?

autumnfjeld replied:
Thanks for those tips!  I moved the check to ETK: https://github.com/Automattic/wp-calypso/pull/51055/commits/95cd894f9ac0fe4119a724102b232592728435c2

autumnfjeld replied:
We will show the NUX modal to mobile users.

simison commented:
Let's keep the old version for mobile and the new version for desktop, if possible. 👍 Next iteration ships to everyone then.

autumnfjeld committed:
Anchor.fm content no longer exists in NUX modal, so we can remove the isPodcastingSite flag and logic

autumnfjeld committed:
WelcomeTour not yet designed for mobile: pass wp_is_mobile() flag to client code in order to hide the Tour component for mobile users 😔

autumnfjeld committed:
On mobile: show the NUX modal, update to use Welcome Tour when mobile designs are ready

autumnfjeld committed:
Remove spurious g

p-jackson committed:
Adjust nux API so old JS clients will work with new PHP code

This is important because during the deployment the server could be
running new PHP code while still serving up old JS files.

I've also updated the logic in `show_editor_onboarding()` method so that
it doesn't force the onboarding to show when it's a mobile device, and
renamed the debugging global to something more aptly named now that the
welcome tour is no longer a test.

It was necessary to reinstate the `showWpcomNuxVariant` part of the
store because otherwise we would always make a request to the `/nux`
endpoint to get the value of `is_mobile_device`, and there would be no
caching.

ramonjd commented:
This LGTM!

**Desktop**

Welcome Tour should show when opening editor ✅ 
go through all slides, minimize, maximize, etc to make sure Tour works ✅ 
Open Tour from Options Menu (aka More Menu) -> "Welcome Guide" ✅ 
Close, then re-open editor and make sure the Welcome Tour does not automatically show ✅ 

**Mobile**

Cleared cache and went through the above steps:
The (existing) NUX Modal should show ✅

autumnfjeld commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/class-wp-rest-wpcom-block-editor-nux-status-controller.php:
```diff
@@ -79,10 +84,18 @@ public function get_nux_status() {
+
+				// These are legacy rest params that can be removed after
+				// we know the new JS files have been deployed.
+				'is_nux_enabled'            => $show_editor_onboarding,
```
Wondering why you think this is a legacy reset param?  The `is_nux_enabled` is important for determining if the user is a new user.   It is a user preference that is set as 'enabled' for new users (as far as I understand).  So this isn't legacy, it will remain and is important so we can determine whether to automatically show the Welcome UI (whether Tour or NUX modal to the user).


(Side note: unfortunately this component is using the `is_nux_enabled` flag to also keep track of the open /closed status of the NUX (or [Tour](https://github.com/Automattic/wp-calypso/blob/578c58b8d687dbc28a59e4d06c2f875a655adfb3/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js#L78)), this is a code cleanup I'd like to do, use a flag purposely for open/close stus and keep the nux flags for the sole purpose of new-user-status')

autumnfjeld replied:
Ah!! I see what you did:
`	$show_editor_onboarding = $this->show_editor_onboarding( $nux_status );`

Should we carry that variable name change throughout the component (as now its a mixup of isNuxEnables and show_editor_onboarding:
https://github.com/Automattic/wp-calypso/blob/8e03ba31060da3b7fe1f6a7d23f4534bbc7bb9d7/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/block-editor-nux.js#L43

https://github.com/Automattic/wp-calypso/blob/3b7743c86c77b3811e512fd351ad63e0b287ddde/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/disable-core-nux.js#L21

https://github.com/Automattic/wp-calypso/blob/3b7743c86c77b3811e512fd351ad63e0b287ddde/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/disable-core-nux.js#L25

And more like that in the NUX modal code and Tour code (e.g. dismiss events).

I think this would then satisfy my itch to have a client side flag that is about 'show' (and not about is_nux_enabled).   I'm just skimming all this so hopefully it all makes sense. :)

p-jackson replied:
Yes that's right, I wanted to rename that variable while I was at it, to make it more about "editor onboarding" and less about "nux". Sorry I hadn't updated the GH thread with any comments yet, they're all buried in the commit message.

I initially wanted to collapse it all into one variable: so something like `editor_onboarding_status` that would be either `false`, `'modal'` or `'tour'`. However I soon realised that the client needs to be able to open the onboarding itself even if `editor_onboarding_status` was `false` when the page loaded, and it needed to have an idea of which variant to load. So left it as two variables.

Yes I think it'd be good to carry the renaming through the rest of the store! I'll do that, but it'd be good to have a call to make sure I'm on the same page as you w.r.t. what sort of language we should be using. I'm thinking
- The umbrella feature is called "editor onboarding"
- The onboarding is either "shown" or "not shown"
- There are two "variants" of the onboarding

I didn't like the idea of using "is mobile" as part of the API. Not sure why, maybe just a taste thing? If felt like it's not high level what we're trying to communicate to the client. The server can make the decision around which variant to show and the client just does what it's told. It also opens up the idea of having `'tour_i1'`, `tour_i2`, etc. if we want to keep iterating.

autumnfjeld replied:
> all buried in the commit message.
A great commit message it is!!  Good example for the rest of us ! :)

> The umbrella feature is called "editor onboarding"
As we chatted about, indeed that is the umbrella feature, of which this Tour (or Guide as core calls it) may just be one part.   Additional help features may be included in Onboarding, besides just the Tour/Guide.

> I didn't like the idea of using "is mobile" as part of the API. Not sure why, maybe just a taste thing?
I first started by passing `show_welcome_tour` then switched to `is_mobile`, was still debating in my head if that was right, so good to hear your opinion and I fully support your changes! :)

p-jackson committed:
Rename "NUX" to "Welcome Guide"

Welcome guide is the term I'm using that covers both the tour and the
modal. They're both "variants" of the "welcome guide".

Core also calls this feature the welcome guide, so it should mean folks
are familiar with the term. When we need to disambiguate our version can
be called the "WPCOM Welcome Guide".

Moved away from calling it "editor onboarding", because that term could
grow to encompass many other features (like tooltips, or guides for
specific blocks). I like the term "welcome" because it gives the
impression that it's the first thing the user sees.

The data store has been renamed to `automattic/wpcom-welcome-guide`
which will invalidate all the localStorage caches in browsers around the
world. These would have been invalidated by the renaming of store
properties.
I've also made sure that the data store is able to deal with the old and
new version of the `/block-editor/nux` API so that it will work across
deploys.

p-jackson commented:
Call with @autumnfjeld inspired even more extensive changes! This time it's just renaming things (but it's a lot of things that needed renaming. The "nux" term is becoming less and less useful since it's a code name that doesn't mean much to us anymore. Ganon will likely work on other "editor onboarding" features over time, of which this is just one.
Which means this PR also addresses #50876 now.

I think _this_ time I'm happy with the changes 😄 (although if I'm feeling inspired I might add some unit tests for the store since a lot has changed, but that shouldn't impact review).

p-jackson commented:
Oh and another wee change, I managed to get rid of the `bypassApi` prop from the action that sets the guide open state. I did this by splitting into two actions:
- One action for toggling open state (which now has the flag for whether it was a manual opening or not)
- One action for loading API response data into the redux store

I think this pattern's a bit more common, having `fetch` action that fires a `blahBlahSuccess` or `blahBlahFailure` action depending on the response back from the server.

autumnfjeld commented:
Tested locally and on my sandbox.

✅  NUX modal shows on mobile & Tour shows on desktop
✅  They respectively show automatically on first editor view and not on second editor view
✅  They are accessible from the Options (aka More) Menu

p-jackson reviewed (approved):
We're good to go!
